### PR TITLE
Better feedback from fingerprint auth and don't clear password field on retry (only on failure)

### PR DESCRIPTION
**Pull request by rickyelopez**
Quick and crude fix targetted at the issue I raised in #608.

### CODE REVIEW

File: src/auth/Fingerprint.hpp [hunk 1]
```diff
@@ -27,6 +27,8 @@ class CFingerprint : public IAuthImplementation {
   private:
     struct SDBUSState {
         std::string                         message = "";
+        std::string                         failure_reason = "";
```
Comment (PaideiaDilemma) [inline src/auth/Fingerprint.hpp:30]: `failureReason`

Comment (rickyelopez) [reply src/auth/Fingerprint.hpp:30]: oh I also forgot to delete `message` after refactoring

Comment (PaideiaDilemma) [reply src/auth/Fingerprint.hpp:30]: Actually while your at it, can you move the strings out of `SDBUSState`?
Either just `m_sMessage, m_sFailureState` and co, or in their own struct. Feel like they don't belong there.

Comment (rickyelopez) [reply src/auth/Fingerprint.hpp:30]: sure that's fair

File: src/auth/Fingerprint.cpp [hunk 2]
```diff
@@ -81,10 +81,14 @@ void CFingerprint::handleInput(const std::string& input) {
 }
 
 std::optional<std::string> CFingerprint::getLastFailText() {
-    return m_sDBUSState.message.empty() ? std::nullopt : std::optional(m_sDBUSState.message);
+    if (!m_sFailureReason.empty() && m_sPrompt.empty())
```
Comment (PaideiaDilemma) [inline src/auth/Fingerprint.cpp:84]: no need to check for prompt i think. same below

File: src/auth/Fingerprint.cpp [hunk 3]
```diff
@@ -153,41 +158,61 @@ void CFingerprint::handleVerifyStatus(const std::string& result, bool done) {
...
             stopVerify();
-            m_sDBUSState.message = "Fingerprint auth disabled (unknown error)";
+            m_sPrompt         = "";
+            m_sFailureReason = "Fingerprint auth disabled (unknown error)";
             break;
         case MATCH_MATCHED:
             stopVerify();
-            m_sDBUSState.message = "";
-            authenticated        = true;
+            m_sPrompt         = "Fingerprint matched!";
```
Comment (PaideiaDilemma) [inline src/auth/Fingerprint.cpp:183]: Wait that is not a "prompt"?
I prompt should indicate what we expect the user to do. `Password: `, `TOTP: ` or in this context smth like `Put your finger on the sensor`.

File: src/auth/Fingerprint.cpp (hunk 2 above)
Comment (rickyelopez) [reply src/auth/Fingerprint.cpp:84]: yeah you're right, originally I wasn't clearing one when setting the other, but I think now it should work the same without this check

File: src/auth/Fingerprint.cpp (hunk 3 above)
Comment (rickyelopez) [reply src/auth/Fingerprint.cpp:183]: fair enough

File: src/auth/Fingerprint.cpp [hunk 4]
```diff
@@ -218,15 +243,20 @@ void CFingerprint::startVerify(bool isRetry) {
...
+                m_sFailureReason = "Fingerprint auth disabled (failed to restart)";
+            }
         } else {
             Debug::log(LOG, "fprint: started verifying");
             if (isRetry) {
                 m_sDBUSState.retries++;
-                m_sDBUSState.message = RETRY_MESSAGE;
-            } else
-                m_sDBUSState.message = m_sFingerprintReady;
+                m_sPrompt        = RETRY_MESSAGE;
```
Comment (rickyelopez) [inline src/auth/Fingerprint.cpp:254]: @PaideiaDilemma just curious, why is this the only message that is predefined as a const?

Comment (PaideiaDilemma) [reply src/auth/Fingerprint.cpp:254]: No particular reason, and its not in style as well (Consts are without `_`). You could remove it and substitute for the string.

**Review by PaideiaDilemma — APPROVED** [review]

**Comment by PaideiaDilemma** [comment]
@rickyelopez needs a wiki mr for `$FPRINTPROMPT` :)

**Comment by rickyelopez** [comment]
> @rickyelopez needs a wiki mr for `$FPRINTPROMPT` :)

Just put up a PR: https://github.com/hyprwm/hyprland-wiki/pull/915

**Review by vaxerski — APPROVED** [review]

**Pull request merged**